pkg/kclient/generator/generators.go

```diff
@@ -0,0 +1,344 @@
+		SelectorLabels: selectorLabels,
+	}
+
+	return generateServiceSpec(serviceSpecParams), nil
+}
+
+// IngressParams struct for function GenerateIngressSpec
+// serviceName is the name of the service for the target reference
+// ingressDomain is the ingress domain to use for the ingress
+// portNumber is the target port of the ingress
+// Path is the path of the ingress
+// TLSSecretName is the target TLS Secret name of the ingress
+type IngressParams struct {
+	ServiceName   string
+	IngressDomain string
+	PortNumber    intstr.IntOrString
+	TLSSecretName string
+	Path          string
+}
+
+// GenerateIngressSpec creates an ingress spec
+func GenerateIngressSpec(ingressParams IngressParams) *extensionsv1.IngressSpec {
+	path := "/"
+	if ingressParams.Path != "" {
+		path = ingressParams.Path
+	}
+	ingressSpec := &extensionsv1.IngressSpec{
+		Rules: []extensionsv1.IngressRule{
+			{
+				Host: ingressParams.IngressDomain,
+				IngressRuleValue: extensionsv1.IngressRuleValue{
+					HTTP: &extensionsv1.HTTPIngressRuleValue{
+						Paths: []extensionsv1.HTTPIngressPath{
+							{
+								Path: path,
+								Backend: extensionsv1.IngressBackend{
+									ServiceName: ingressParams.ServiceName,
+									ServicePort: ingressParams.PortNumber,
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+	secretNameLength := len(ingressParams.TLSSecretName)
+	if secretNameLength != 0 {
+		ingressSpec.TLS = []extensionsv1.IngressTLS{
+			{
+				Hosts: []string{
+					ingressParams.IngressDomain,
+				},
+				SecretName: ingressParams.TLSSecretName,
+			},
+		}
+	}
+
+	return ingressSpec
+}
+
+// RouteParams struct for function GenerateRouteSpec
+// serviceName is the name of the service for the target reference
+// portNumber is the target port of the ingress
+// Path is the path of the route
+type RouteParams struct {
+	ServiceName string
+	PortNumber  intstr.IntOrString
+	Path        string
+	Secure      bool
+}
+
+// GenerateRouteSpec creates a route spec
+func GenerateRouteSpec(routeParams RouteParams) *routev1.RouteSpec {
+	routePath := "/"
+	if routeParams.Path != "" {
+		routePath = routeParams.Path
+	}
+	routeSpec := &routev1.RouteSpec{
+		To: routev1.RouteTargetReference{
+			Kind: "Service",
+			Name: routeParams.ServiceName,
+		},
+		Port: &routev1.RoutePort{
+			TargetPort: routeParams.PortNumber,
+		},
+		Path: routePath,
+	}
+
+	if routeParams.Secure {
+		routeSpec.TLS = &routev1.TLSConfig{
+			Termination:                   routev1.TLSTerminationEdge,
+			InsecureEdgeTerminationPolicy: routev1.InsecureEdgeTerminationPolicyRedirect,
+		}
+	}
+
+	return routeSpec
+}
+
+// GenerateOwnerReference genertes an ownerReference  from the deployment which can then be set as
+// owner for various Kubernetes objects and ensure that when the owner object is deleted from the
+// cluster, all other objects are automatically removed by Kubernetes garbage collector
+func GenerateOwnerReference(deployment *appsv1.Deployment) metav1.OwnerReference {
```

Comment:
we should also have a similar function for deploymentConfigs

Reply:
Do we really need it? Because only s2i maybe deploymentConfigs but devfiles would most probably always be deployments

Reply:
Someone might decide to use deployment configs with devfile in future. Anyways, we can leave this out of the generator for now.